crates/types/src/traits/block_contents.rs

```diff
@@ -47,20 +46,27 @@ pub trait BlockPayload:
     type Transaction: Transaction;
 
     /// Data created during block building which feeds into the block header
-    type Metadata;
+    type Metadata: Clone + Debug + Eq + Hash + Send + Sync;
+
+    /// Encoded payload.
+    type Encode<'a>: 'a + Iterator<Item = u8> + AsRef<[u8]> + Send;
 
     /// Build a payload and associated metadata with the transactions.
-    fn build(transactions: impl IntoIterator<Item = Self::Transaction>) -> (Self, Self::Metadata);
+    fn from_transactions(
+        transactions: impl IntoIterator<Item = Self::Transaction>,
+    ) -> (Self, Self::Metadata);
 
-    /// Encode the payload
-    fn encode(&self) -> Vec<u8>;
+    /// Build a payload with the encoded transaction bytes and metadata.
+    fn from_bytes(encoded_transactions: Self::Encode<'_>, metadata: Self::Metadata) -> Self;
```

Comment:
`Self::Encode` may be a problem here. This means the method can only be called with a specific type of iterator, a type which can only be constructed by calling `encode`. What if we want to serialize the results of `encode`, send it over the network, receive it as a `Vec<u8>`, and then decode that? I think this function should be generic:
```suggestion
    fn from_bytes<I>(encoded_transactions: I, metadata: Self::Metadata) -> Self
    where I: Iterator<Item = u8>;
```